HEEDLS-916 email verification registration

livzorn opened the pull request:
### JIRA link
[_HEEDLS-916_](https://softwiretech.atlassian.net/browse/HEEDLS-916)

### Description
Made the following changes:
1. Made sure in all self-registration journeys, the user's primary and centre email are marked as unverified
2. Made sure in registration by admin journeys, the user's centre email is marked as verified
3. Added a warning to self-registration confirmation pages about verifying your email (I've modified this from the wireframe, making it more similar to the one showed on the My account page, for simplicity of code and also to give the user's more info. I've also been playing around with the wording and bolding emails. Please review screenshots and let me know what you think!)
4. Fixed a bug with the service filter, the controller and action name had not been changed. And added this service filter to the internal admin registration journey (it was already on the internal delegate registration journey).
5. Left TODO comments for 915, for where to add a call to SendVerificationLink and IsEmailVerifiedForUser
6. Fixed a cancel link and a typo in an error message

### Screenshots
External delegate registration
![localhost_44363_Register_Confirmation_centreId=222 candidateNumber=OZ37 approved=True unverifiedPrimaryEmail=olivia@zornA com CentreSpecificEmail=olivia@zornB com](https://user-images.githubusercontent.com/70278044/182863511-748cb27f-8c23-4ea6-b94a-4df023cb0436.png)

Internal delegate registration 
![localhost_44363_RegisterAtNewCentre_Confirmation_candidateNumber=OZ38 approved=True userHasAdminAccountAtCentre=False centreId=279](https://user-images.githubusercontent.com/70278044/182863545-99bf7e54-7a30-4bb7-a0c8-3db2a76f71c0.png)

External admin registration
![localhost_44363_RegisterAdmin_Confirmation_primaryEmail=olivia@test800 com centreId=80 centreSpecificEmail=olivia@test8000 com](https://user-images.githubusercontent.com/70278044/182863653-7a0b0354-d79b-4816-a9a1-882d0cf043e1.png)

Internal admin registration
![localhost_44363_RegisterInternalAdmin_Confirmation_centreId=80](https://user-images.githubusercontent.com/70278044/182863600-44e1d2eb-95b5-4c00-b80c-eb4ca923e48e.png)

Tablet example
![localhost_44363_Register_Confirmation_centreId=222 candidateNumber=OZ37 approved=True unverifiedPrimaryEmail=olivia@zornA com CentreSpecificEmail=olivia@zornB com (1)](https://user-images.githubusercontent.com/70278044/182863736-8ee64472-6125-4ddf-bf4e-e385c6763ee5.png)

Mobile example
![localhost_44363_Register_Confirmation_centreId=222 candidateNumber=OZ37 approved=True unverifiedPrimaryEmail=olivia@zornA com CentreSpecificEmail=olivia@zornB com (2)](https://user-images.githubusercontent.com/70278044/182863745-f9b2ebe7-fdc1-46fd-bbd3-3d967bcd66ff.png)


-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [ ] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
  - [doc_1_here](link_1_here)
- [x] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

SteveJacksonSoft commented on DigitalLearningSolutions.Data/DataServices/RegistrationDataService.cs:
```diff
@@ -281,16 +290,21 @@ OUTPUT Inserted.ID
             if (!string.IsNullOrWhiteSpace(centreSpecificEmail))
             {
+                // TODO: 915 - replace false with call to IsEmailVerifiedForUser
+                var emailIsVerifiedForUser = false;
+                var emailShouldBeVerified = emailIsVerifiedForUser || !isSelfRegistered;
```
Because this is a data service, I'd like it to be dumb as possible and pull logic into the Registration Service. Can just pass `bool emailIsVerified` in as a param?

livzorn replied:
Moved the logic into the service methods that call this, it caused quite a lot of ripple effects since it's used in a bunch of methods

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Views/RegisterAtNewCentre/Confirmation.cshtml:
```diff
@@ -21,12 +23,20 @@
       <p>Your delegate number is <span class="nhsuk-u-font-weight-bold">@Model.CandidateNumber</span></p>
     </div>
 
+    <partial name="_VerifyEmailWarningForRegistrationConfirmationPage" model="@verifyEmailWarningViewModel" />
```
Looks like this will always appear.

SteveJacksonSoft replied:
Shouldn't assume that we'll have to show this on internal journeys.

livzorn replied:
added a conditional

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Controllers/Register/RegisterAtNewCentreController.cs:
```diff
@@ -266,31 +271,35 @@ public async Task<IActionResult> SummaryPost()
-                centreId
+                userHasAdminAccountAtCentre,
+                centreId,
+                unverifiedPrimaryEmail,
```
This is an internal registration journey, so we only need to show the centre email if at all.

SteveJacksonSoft replied:
Similar on RegisterInternalAdminController

livzorn replied:
removed in both journeys

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Controllers/Register/RegisterAtNewCentreController.cs:
```diff
@@ -266,31 +271,35 @@ public async Task<IActionResult> SummaryPost()
+
+            var (unverifiedPrimaryEmail, unverifiedCentreEmails) =
+                userService.GetUnverifiedEmailsForUser(userId);
+            var (_, centreName, unverifiedCentreEmail) = unverifiedCentreEmails.First(uce => uce.centreId == centreId);
```
*SingleOrDefault - may turn out to be the same as a verified centre email / primary email.

SteveJacksonSoft replied:
Similar on RegisterInternalAdminController

livzorn replied:
changed in both journeys

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Controllers/Register/RegisterController.cs:
```diff
@@ -311,24 +321,32 @@ public async Task<IActionResult> Summary(SummaryViewModel model)
 
-            var approved = (bool)approvedNullable;
-            var centreId = (int)centreIdNullable;
+            var centreIdForContactInformation = approved ? null : centreId;
```
I'd expect this line to be in the view...

livzorn replied:
This is just the centreId used for getting the contact info, looks like the actual info is retrieved from the db in the CentreContactInfoViewComponent

SteveJacksonSoft replied:
What I mean is that we may as well pass the centreId to the view. The controller shouldn't be deciding when the view needs the centreId - that's the view's responsibility.

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Controllers/Register/RegisterInternalAdminController.cs:
```diff
@@ -125,13 +130,28 @@ await featureManager.IsEnabledAsync("RefactoredTrackingSystem")
-            return RedirectToAction("Confirmation");
+            if (model.CentreSpecificEmail != null)
+            {
+                // TODO: 915 - Send verification link to centre email
```
*if not already verified

SteveJacksonSoft commented on DigitalLearningSolutions.Web/DigitalLearningSolutions.Web.csproj:
```diff
@@ -16,129 +16,133 @@
     </PropertyGroup>
```
Any idea why this has been modified?

livzorn replied:
Hmm no idea, I'll change it back.

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Models/DelegateRegistrationData.cs:
```diff
@@ -1,16 +1,23 @@
+            int? centreId,
+            int? supervisorDelegateId = null,
+            string? primaryEmail = null,
+            DateTime? primaryEmailVerified = null
```
This doesn't look like it's every getting used...

livzorn replied:
removed

SteveJacksonSoft commented on DigitalLearningSolutions.Web/ViewModels/Register/InternalAdminInformationViewModel.cs:
```diff
@@ -1,7 +1,10 @@
+    public class InternalAdminInformationViewModel : InternalPersonalInformationViewModel
+    {
+        public string PrimaryEmail { get; set; } = null!;
+        public DateTime? PrimaryEmailVerified { get; set; } = null;
```
Unused I'm told.

livzorn replied:
removed

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Views/RegisterInternalAdmin/Confirmation.cshtml:
```diff
@@ -1,18 +1,25 @@
     </p>
 
-    <vc:action-link asp-controller="Login" asp-action="ChooseACentre" link-text="Switch centre" />
+    <partial name="_VerifyEmailWarningForRegistrationConfirmationPage" model="@verifyEmailWarningViewModel" />
```
Shouldn't assume that we'll have to show this on internal journeys.

livzorn replied:
added a conditional

SteveJacksonSoft commented:
Appearance of callouts looks great!

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Views/Register/Confirmation.cshtml:
```diff
@@ -35,6 +43,8 @@
       </div>
       <vc:centre-contact-info centre-id="@Model.CentreId.GetValueOrDefault()" />
+    } else {
+      <vc:action-link link-text="Log in" asp-controller="Login" asp-action="Index" />
```
Kevin wants this, so good job.

livzorn commented on DigitalLearningSolutions.Web/ViewModels/Login/ChooseACentreViewModel.cs:
```diff
@@ -1,26 +1,27 @@
             string? returnUrl,
-            bool primaryEmailIsVerified,
-            List<(int centreId, string centreName, string? centreSpecificEmail)> unverifiedCentreEmails
+            string? unverifiedPrimaryEmail,
```
Hmm, think that was a mistake. Undone.

livzorn commented on DigitalLearningSolutions.Web/Controllers/MyAccountController.cs:
```diff
@@ -79,7 +79,8 @@ public IActionResult Index(DlsSubApplication dlsSubApplication)
                 : new List<(int centreId, string centreName, string? centreSpecificEmail)>();
 
-            var (_, unverifiedCentreEmails) = userService.GetUnverifiedEmailsForUser(userEntity.UserAccount.Id);
+            var (unverifiedPrimaryEmail, unverifiedCentreEmails) =
```
Nope, removed

livzorn commented on DigitalLearningSolutions.Web/Views/Shared/Components/VerifyEmailWarningIfAppropriate/Default.cshtml:
```diff
@@ -8,28 +8,20 @@
+    } else if (Model.IsMyAccountPage) {
+      <partial name="_VerifyEmailTextAndResendButtonForMyAccount" model="@Model" />
+    } else if (Model.IsChooseACentrePage) {
+      <partial name="_VerifyEmailTextForChooseACentre" model="@Model" />
```
Removed the view components and made partials instead

livzorn commented on DigitalLearningSolutions.Web/ViewModels/Login/ChooseACentreViewModel.cs:
```diff
@@ -1,26 +1,28 @@
         public string? ReturnUrl { get; set; }
-        public bool PrimaryEmailIsVerified { get; set; }
-        public List<(int centreId, string centreName, string? centreSpecificEmail)> UnverifiedCentreEmails { get; set; }
+        public string? UnverifiedPrimaryEmail { get; set; }
```
I like that, changed

livzorn commented on DigitalLearningSolutions.Web/Views/Register/PersonalInformation.cshtml:
```diff
@@ -99,5 +94,9 @@
     </a>
+
+    <div class="nhsuk-u-margin-top-3">
+      <vc:cancel-link asp-controller="Home" asp-action="Index" />
```
Kevin brought this up in hee-dls-refactor, so I figured I'd fix it in this ticket

livzorn commented on DigitalLearningSolutions.Web/Views/RegisterAdmin/PersonalInformation.cshtml:
```diff
@@ -75,5 +73,9 @@
     </a>
+
+    <div class="nhsuk-u-margin-top-3">
+      <vc:cancel-link asp-controller="Home" asp-action="Index" />
```
Same, Kevin noticed this during show and tell and wanted it moved

livzorn commented on DigitalLearningSolutions.Web/Views/RegisterInternalAdmin/Confirmation.cshtml:
```diff
@@ -1,18 +1,30 @@
 <div class="nhsuk-grid-row">
   <div class="nhsuk-grid-column-full">
-    <h1 class="nhsuk-heading-xl">Admin registration complete</h1>
+    <h1 id="page-heading" class="nhsuk-heading-xl">Admin registration complete</h1>
```
I saw this was present on the headings of the other confirmation pages, it removes some extra space between the heading and the text below. All the pages have consistent spacing.

livzorn commented on DigitalLearningSolutions.Web/Views/RegisterAtNewCentre/Confirmation.cshtml:
```diff
@@ -21,12 +22,25 @@
       </p>
       <partial name="_LogoutForm" model="@ButtonType.Secondary" />
-    } else if (Model.Approved) {
+    } else if (Model.Approved && emailsAreVerified) {
```
I've cleaned this up a bit. The only thing I'm not certain of is the priority of the action links/buttons. I've put them in this order but feel free to disagree:
1. Logout button (For when the user has an admin account at that centre. They should be auto-approved and they might still have an unverified email, but the verify email warning tells them to visit the My Account page if they need to resend the verification link, so prompting them to log out seems more important)
2. Switch centre button (For when the account is approved and the email is verified)
3. Centre contact info action link (For when the account if not approved. The email might still be unverified but the verify email warning tells them to visit the My Account page if they need to resend the verification link, so showing them where to find centre contact info seems more important as they might not otherwise know where to find it.)
4. View My Account action link (For if the email is unverified but all else is well - account approved, no admin account at the centre.)

SteveJacksonSoft replied:
So actually, I think we should have the switch centre button for the admin as well. I don't think we need to treat them separately from other auto-approved users. All they need to do is re-log-into the current centre. If we log them out then they have to enter credentials again which'll be a tad annoying for them. Can we just show them the standard auto-approved text?

It feels strange to send them to MyAccount if the email is unverified - they'll still be logged into whichever centre they're currently logged into and so they can't even see the centre email they need to verify. I think switch-centre is probably still suitable for them - it's where they'll want to go after verifying their email anyway.

livzorn replied:
Sure, I've merged the HasAdminAccount condition with the one that shows the switch centre button

livzorn commented:
Adjusted the My account page warning and the Verify Email page to use the same view component for listing the unverified emails. Also adjusted these and the Registration confirmation warnings to account for an unverified email being used by multiple centres or the unverified primary email also being used as a centre email. 
![localhost_44363_MyAccount (5)](https://user-images.githubusercontent.com/70278044/184318761-9590b412-7160-4e1e-a52d-20c5f25f7731.png)
![localhost_44363_MyAccount (6)](https://user-images.githubusercontent.com/70278044/184318770-390989b7-920c-4dba-9fd9-10eda0cafd36.png)
![localhost_44363_MyAccount (7)](https://user-images.githubusercontent.com/70278044/184318780-b7cdee4c-fb0e-4230-ad78-ac0bdb8f6c89.png)
![localhost_44363_Register_Confirmation_centreId=409 candidateNumber=OZ48 approved=True unverifiedPrimaryEmail=olivia@sameEmail com CentreSpecificEmail=olivia@sameEmail com](https://user-images.githubusercontent.com/70278044/184318788-4ae7133d-6cd6-45f8-adf1-1a8df8811b0b.png)
![localhost_44363_VerifyYourEmail_EmailNotVerified](https://user-images.githubusercontent.com/70278044/184318805-8e4c56a8-5c39-4126-bf1d-475292bdc55d.png)

livzorn commented on DigitalLearningSolutions.Web.Tests/Services/RegistrationServiceTests.cs:
```diff
@@ -1629,19 +1722,19 @@ private void GivenAdminsToNotifyHaveEmails(IEnumerable<string> adminEmails)
         }
 
-        private void RegistrationRequiresApprovalEmailMustHaveBeenSentTo(NotificationRecipient notificationRecipient)
+        private void RegistrationRequiresApprovalEmailMustHaveBeenSentTo(string recipientEmail, int numberOfTimes = 1)
```
Noticed that this method was unused, so I added it in the tests above where it was intended to go

SteveJacksonSoft replied:
Is it still unused?

livzorn replied:
No, it's now used in the tests above

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Controllers/Register/RegisterInternalAdminController.cs:
```diff
@@ -143,13 +156,37 @@ await featureManager.IsEnabledAsync("RefactoredTrackingSystem")
+                !emailVerificationService.AccountEmailIsVerifiedForUser(userId, model.CentreSpecificEmail))
+            {
+                var userEntity = userService.GetUserById(userId);
+                var unverifiedEmails = PossibleEmailUpdateHelper.GetSingleUnverifiedEmail(model.CentreSpecificEmail);
```
Don't think you need this. I'm pretty sure you're just wrapping the email here and unwrapping it a few lines later...

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Controllers/Register/RegisterController.cs:
```diff
@@ -278,11 +290,25 @@ public async Task<IActionResult> Summary(SummaryViewModel model)
                         data.SupervisorDelegateId
                     );
 
+                CreateEmailVerificationHashesAndSendVerificationEmails(
```
This doesn't feel like logic that needs to be in the controller. I think it should be merged into registrationService.CreateDelegateAccountForNewUser, and I think we can rename that method to simply "RegisterDelegateForNewUser" since this is the only place we use it, and it does really do a bit more than just creating a delegate account..

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Controllers/Register/RegisterAtNewCentreController.cs:
```diff
@@ -245,12 +253,25 @@ public async Task<IActionResult> SummaryPost()
+                    !emailVerificationService.AccountEmailIsVerifiedForUser(userId, data.CentreSpecificEmail))
+                {
+                    var userEntity = userService.GetUserById(userId);
+                    var unverifiedEmails = PossibleEmailUpdateHelper.GetSingleUnverifiedEmail(data.CentreSpecificEmail);
```
Don't think this is necessary.

SteveJacksonSoft commented on DigitalLearningSolutions.Data/DataServices/RegistrationDataService.cs:
```diff
@@ -14,7 +14,8 @@ public interface IRegistrationDataService
             DelegateRegistrationModel delegateRegistrationModel,
-            bool registerJourneyContainsTermsAndConditions
+            bool registerJourneyContainsTermsAndConditions,
+            bool isRegisteredByAdmin
```
I think you've got a tendency to pass context information ("On my account page", "registered by admin") as parameters into lower level methods, then have a single low level method that branches to handle different contexts.
As a result:
    * The behaviour for a particular context is spread down the stack of methods rather than being in one place
    * The low level method depends on all the contexts, so may have to change if any of the business logic for any of the contexts changes.
    * Any new scenarios which want to use that low level method might have to modify that low level method to extend the set of context parameters passed into the method (because it might want to do something slightly different from what's currently there) - this runs the risk of breaking existing code that calls that method.
    
A better approach is to make everything depend on abstractions (dependency inversion) - decide on a thing that needs to happen (e.g. your low level method does a thing

SteveJacksonSoft replied:
Oh, apparently this method isn't used outside of RegistrationService either - can you remove it from the interface?

SteveJacksonSoft replied:
Can we rename `isRegisteredByAdmin` to `shouldAssumeEmailVerified`? or something of that ilk. Then we can look at the RegisterByAdmin methods (e.g. `RegisterDelegateByCentre()`) and see that we're automatically verifying the delegate email in that case, rather than having to follow the call stack into `RegistrationDataService` to see what we do if `isRegisteredByAdmin`.

livzorn replied:
Sure! renamed

livzorn commented on DigitalLearningSolutions.Data/DataServices/RegistrationDataService.cs:
```diff
@@ -306,12 +312,16 @@ OUTPUT Inserted.ID
             PossibleEmailUpdate? possibleEmailUpdate,
-            IDbTransaction transaction
+            IDbTransaction transaction,
+            DateTime? currentTime = null
```
This method is sometimes called in other methods that have already calculated the current time to update other db values. I figured we could just pass it in rather than calculating the current time more than once. It represents the time to update the EmailVerified to, which will be near enough the current time. Is it better to rename it, or just calculate the time again here?

SteveJacksonSoft replied:
I don't think it'll give a noticeable performance boost (DateTime.UtcNow isn't that expensive), so I think it's more worthwhile to make the code simpler and check it every time.

livzorn commented on DigitalLearningSolutions.Web/Helpers/PossibleEmailUpdateHelper.cs:
```diff
@@ -0,0 +1,54 @@
+
+    public static class PossibleEmailUpdateHelper
+    {
+        public static List<PossibleEmailUpdate> GetUnverifiedPrimaryAndCentreEmails(
```
No need for this in the end, deleted it

livzorn commented on DigitalLearningSolutions.Web/Controllers/LoginController.cs:
```diff
@@ -126,7 +126,7 @@ public IActionResult ChooseACentre(DlsSubApplication dlsSubApplication, string?
                     .ThenBy(account => account.CentreName).ToList(),
                 returnUrl,
-                userEntity.UserAccount.EmailVerified.HasValue,
+                unverifiedPrimaryEmail == null,
```
Can't remember why I changed it, but they should be the same, so I've changed it back

SteveJacksonSoft commented on DigitalLearningSolutions.Data.Tests/TestHelpers/UserTestHelper.cs:
```diff
@@ -160,6 +161,19 @@ public static class UserTestHelper
+        public static UserEntity GetDefaultUserEntity(
+            int userId = 2,
+            string primaryEmail = "primary@email.com",
+            string? centreSpecificEmail = "centre@email.com"
```
You're not doing anything with the centre email here, which looks a bit fishy.

SteveJacksonSoft commented on DigitalLearningSolutions.Web.Tests/TestHelpers/ListTestHelper.cs:
```diff
@@ -0,0 +1,21 @@
+            List<string> list2
+        )
+        {
+            if (list1.Count != list2.Count)
+            {
+                return false;
+            }
+
+            return list1.All(list2.Contains);
```
This doesn't quite do what you want it to do - it'll return true if list 1 is {"1", "1"} and list 2 is {"1", "2"}.
I think you want the first option from https://www.techiedelight.com/check-whether-two-lists-have-the-same-items-in-csharp/

livzorn commented on DigitalLearningSolutions.Web/Controllers/Register/RegisterAtNewCentreController.cs:
```diff
@@ -256,12 +257,11 @@ public async Task<IActionResult> SummaryPost()
                 {
-                    var userEntity = userService.GetUserById(userId);
-                    var unverifiedEmails = PossibleEmailUpdateHelper.GetSingleUnverifiedEmail(data.CentreSpecificEmail);
+                    var userAccount = userService.GetUserByEmailAddress(User.GetUserPrimaryEmailKnownNotNull());
```
Cool, done that!

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Views/RegisterAtNewCentre/Confirmation.cshtml:
```diff
@@ -31,12 +31,7 @@
-      </p>
-      <partial name="_LogoutForm" model="@ButtonType.Secondary" />
-    } else if (Model.Approved && centreEmailIsVerified) {
+    @if (Model.HasAdminAccountAtCentre || Model.Approved && centreEmailIsVerified) {
```
HasAdminAccountAtCentre => Approved so I'd say get rid of `Model.HasAdminAccountAtCentre ||`.

livzorn replied:
But we still want to show admins the switch centre button if their email is unverified right? So it would be Model.HasAdminAccountAtCentre || (Model.Approved && centreEmailIsVerified)

SteveJacksonSoft replied:
We'll show Switch Centre to all approved registrants.

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Views/RegisterAtNewCentre/Confirmation.cshtml:
```diff
@@ -21,12 +22,25 @@
+    @if (emailsAreVerified) {
+      <vc:action-link asp-controller="Login" asp-action="ChooseACentre" link-text="Switch centre" />
+    } else {
+      <vc:action-link asp-controller="MyAccount" asp-action="Index" link-text="View My account" />
```
Let's always show approved users switch-centre

SteveJacksonSoft commented on DigitalLearningSolutions.Web/Views/RegisterInternalAdmin/Confirmation.cshtml:
```diff
@@ -1,18 +1,30 @@
+      unverified-centre-emails="@Model.UnverifiedCentreEmails" />
+
+    @if (emailsAreVerified) {
+      <vc:action-link asp-controller="Login" asp-action="ChooseACentre" link-text="Switch centre" />
+    } else {
+      <vc:action-link asp-controller="MyAccount" asp-action="Index" link-text="View My account" />
```
Let's always show users switch-centre
